admiral/pkg/controller/admiral/job.go

```diff
@@ -0,0 +1,273 @@
+
+	jce, ok := p.cache[identity]
+	if ok {
+		jobFromNamespace, ok := jce.Jobs[job.Namespace]
+		if ok {
+			return jobFromNamespace.Status, nil
+		}
+	}
+
+	return common.NotProcessed, nil
+}
+
+func (p *jobCache) UpdateJobProcessStatus(job *v12.Job, status string) error {
+	defer p.mutex.Unlock()
+	p.mutex.Lock()
+
+	identity := common.GetGlobalIdentifier(job.Annotations, job.Labels)
+
+	jce, ok := p.cache[identity]
+	if ok {
+		jobFromNamespace, ok := jce.Jobs[job.Namespace]
+		if ok {
+			jobFromNamespace.Status = status
+			p.cache[jce.Identity] = jce
+			return nil
+		} else {
+			newJob := p.getK8sObjectFromJob(job)
+			newJob.Status = status
+			jce.Jobs[job.Namespace] = newJob
+			p.cache[jce.Identity] = jce
+			return nil
+		}
+	}
+
+	return fmt.Errorf(LogCacheFormat, "UpdateStatus", "Job",
+		job.Name, job.Namespace, "", "nothing to update, job not found in cache")
+}
+
+func (p *JobController) DoesGenerationMatch(ctxLogger *log.Entry, obj interface{}, oldObj interface{}) (bool, error) {
+	if !common.DoGenerationCheck() {
+		ctxLogger.Debugf(ControllerLogFormat, "DoesGenerationMatch", "",
+			fmt.Sprintf("generation check is disabled"))
+		return false, nil
+	}
+	jobNew, ok := obj.(*v12.Job)
+	if !ok {
+		return false, fmt.Errorf("type assertion failed, %v is not of type *Job", obj)
+	}
+	jobOld, ok := oldObj.(*v12.Job)
+	if !ok {
+		return false, fmt.Errorf("type assertion failed, %v is not of type *Job", oldObj)
+	}
+	if jobNew.Generation == jobOld.Generation {
+		ctxLogger.Infof(ControllerLogFormat, "DoesGenerationMatch", "",
+			fmt.Sprintf("old and new generation matched for job %s", jobNew.Name))
+		return true, nil
+	}
+	return false, nil
+}
+
+func NewJobController(stopCh <-chan struct{}, handler ClientDiscoveryHandler, config *rest.Config, resyncPeriod time.Duration, clientLoader loader.ClientLoader) (*JobController, error) {
+
+	jobController := JobController{}
+	jobController.JobHandler = handler
+
+	var err error
+
+	jobController.K8sClient, err = clientLoader.LoadKubeClientFromConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create dependency controller k8s client: %v", err)
+	}
+
+	jobController.informer = v1.NewJobInformer(
+		jobController.K8sClient,
+		meta_v1.NamespaceAll,
+		resyncPeriod,
+		cache.Indexers{},
+	)
+
+	jobController.Cache = NewJobCache()
+
+	NewController("job-ctrl", config.Host, stopCh, &jobController, jobController.informer)
+
+	return &jobController, nil
+}
+
+func (d *JobController) Added(ctx context.Context, obj interface{}) error {
+	return addUpdateJob(d, ctx, obj)
+}
+
+func (d *JobController) Updated(ctx context.Context, obj interface{}, oldObj interface{}) error {
+	//Not Required, to be handled via off boarding
+	return nil
+}
+
+func addUpdateJob(j *JobController, ctx context.Context, obj interface{}) error {
+	job, ok := obj.(*v12.Job)
+	if !ok {
+		return nil
+	}
+	if !common.ShouldIgnore(job.Annotations, job.Labels) {
```

Comment:
Jobs created by cronjobs would also start getting watched by this controller. In the `ShouldIgnore` function we should also ignore jobs which have an `ownerReferences` section. WDYT?

Reply:
All mesh annotations go under job spec in a cron job, so it should be fine to support this at the job spec/job level.

Reply:
Got it, makes sense.